clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/core/agent/CleanupPendingOnDemandCachesAgent.java

```diff
@@ -0,0 +1,117 @@
+package com.netflix.spinnaker.clouddriver.core.agent;
+
+import com.netflix.spinnaker.cats.agent.RunnableAgent;
+import com.netflix.spinnaker.cats.module.CatsModule;
+import com.netflix.spinnaker.cats.redis.JedisSource;
+import com.netflix.spinnaker.clouddriver.cache.CustomScheduledAgent;
+import com.netflix.spinnaker.clouddriver.core.provider.CoreProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.ApplicationContext;
+import redis.clients.jedis.Jedis;
+import redis.clients.jedis.Pipeline;
+import redis.clients.jedis.ScanParams;
+import redis.clients.jedis.ScanResult;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+public class CleanupPendingOnDemandCachesAgent implements RunnableAgent, CustomScheduledAgent {
+  private static final Logger log = LoggerFactory.getLogger(CleanupPendingOnDemandCachesAgent.class);
+
+  private static final long DEFAULT_POLL_INTERVAL_MILLIS = TimeUnit.MINUTES.toMillis(30);
+  private static final long DEFAULT_TIMEOUT_MILLIS = TimeUnit.MINUTES.toMillis(5);
+
+  private final JedisSource jedisSource;
+  private final ApplicationContext applicationContext;
+  private final long pollIntervalMillis;
+  private final long timeoutMillis;
+
+  public CleanupPendingOnDemandCachesAgent(JedisSource jedisSource,
+                                           ApplicationContext applicationContext) {
+    this(jedisSource, applicationContext, DEFAULT_POLL_INTERVAL_MILLIS, DEFAULT_TIMEOUT_MILLIS);
+  }
+
+  private CleanupPendingOnDemandCachesAgent(JedisSource jedisSource,
+                                            ApplicationContext applicationContext,
+                                            long pollIntervalMillis,
+                                            long timeoutMillis) {
+    this.jedisSource = jedisSource;
+    this.applicationContext = applicationContext;
+    this.pollIntervalMillis = pollIntervalMillis;
+    this.timeoutMillis = timeoutMillis;
+  }
+
+  @Override
+  public String getAgentType() {
+    return CleanupPendingOnDemandCachesAgent.class.getSimpleName();
+  }
+
+  @Override
+  public String getProviderName() {
+    return CoreProvider.PROVIDER_NAME;
+  }
+
+  @Override
+  public void run() {
+    getCatsModule().getProviderRegistry().getProviders().forEach(provider -> {
```

Comment:
Specifically this aims to cleanup the onDemand:members sets for Aws/AwsInfrastructure/Titus.